cmd/containerd-shim-runhcs-v1/exec_hcs.go

```diff
@@ -290,7 +290,32 @@ func (he *hcsExec) Kill(ctx context.Context, signal uint32) error {
 		}
 		var delivered bool
 		if supported && options != nil {
-			delivered, err = he.p.Process.Signal(ctx, options)
+			if he.isWCOW {
+				// Servercore images block on signaling and wait until the target process
+				// is terminated to return to the caller. This causes issues when graceful
+				// termination of containers is requested (Bug36689012).
+				// To fix this, we deliver the signal to the target process in a separate background
+				// thread so that the caller can wait for the desired timeout before sending
+				// a SIGKILL to the process.
+				// TODO: We can get rid of these changes once the fix to support graceful termination is
+				// made in windows.
+				go func() {
```

Comment:
I am thinking about if we should leave this "make a background goroutine and return successfully" part here, or if it should really live in the hcs/process.go implementation, as it seems specific to that implementation's handling of signals. Have you thought about this?

I'm also wondering what is the behavior today (pre-fix) for hypervisor-isolated Windows containers. Do they have the same behavior where we can only make one Kill call at a time due to the GCS serializing calls (like HCS does)?

Reply:
I have considered that in the past but thought of putting it in the caller for better readability to understand what's being done in the Kill() call. I could move it out to the Signals() call.

GCS doesn't serialize calls and windows implementation of GCS doesn't serialize either (was able to confirm this from Kris as well). On giving this a second thought, I think the chunk of code for creating a new compute system handle in hcs/process.go could be restricted for only wcow process isolated cases, right?

Reply:
This chunk of code was added here to fix graceful termination issue seen on wcow hyperV cases too. GCS doesn't serialize signals like HCS